scalableminds/webknossos
# Improve group expansion handling in the tree view

- This is a proposal to somewhat improve the group expansion handling in the tree view. In my opinion the current state which auto-expands all groups when loading the tracing is not ideal. When working with larger tracings, the immediate first step after opening the tracing is always to collapse all groups to be able to get an overview of what is contained in that tracing. This "annoyance" has become even more prominent with the recent change that doesn't render the tree tab when it's not visible, because when switching to another tab and back to the tree tab all groups are expanded again. Therefore, I switched the default to be that all groups but the root group are collapsed when opening the tracing.
All parent groups of the active tree or group are expanded, nevertheless, which doesn't change with this PR.
- I've also fixed a bug where the auto-expansion of the parent groups of the active tree/group was not properly persisted in the state. This bug led to very annoying errors where a group was suddenly collapsed when a single tree was made visible/invisible (see steps to test).
- Also add an indeterminate state to groups to indicate if some but not all trees/groups inside of it are visible.
- Do not show the visibility checkbox for groups that don't contain a tree

This PR doesn't persist the expansion state across reloads - this is another issue.

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- Test that the group expansion state of auto-expanded groups is persisted correctly:
  - Open a tracing which has multiple trees inside a group.
  - Collapse all groups but the root group.
  - Activate another tree by shift-clicking it in the tracing view (the parent groups of that tree should be expanded).
  - Change the visibility of another tree within the same group.
  -> The group expansion state should not change (on master the group was collapsed now).
- When opening a tracing, only the parent groups of the active tree/group should be expanded.
- Check that groups where not all but some children are visible show up as indeterminate.

### Issues:
- fixes some annoyances when working with large tracings which contain many groups and trees
- fixes #4788
- fixes #4858

------
